modules/cloudsql/main.tf

```diff
@@ -0,0 +1,106 @@
+/**
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#--------#
+# Locals #
+#--------#
+locals {
+  random_hash     = var.suffix
+  cloudsql_name   = "forseti-server-db-${local.random_hash}"
+  network_project = var.network_project != "" ? var.network_project : var.project_id
+}
+
+#------------------------------------#
+# Forseti Private SQL Database Setup #
+#------------------------------------#
+
+data "google_compute_network" "cloudsql_private_network" {
+  name    = "${var.network}"
+  project = "${local.network_project}"
+}
+
+resource "google_project_service" "service_networking" {
+  count              = var.cloudsql_private ? 1 : 0
+  project            = var.project_id
+  service            = "servicenetworking.googleapis.com"
+  disable_on_destroy = "false"
+}
+
+resource "google_compute_global_address" "private_ip_address" {
+  count         = var.cloudsql_private ? 1 : 0
+  project       = var.project_id
+  name          = "private-ip-address"
+  purpose       = "VPC_PEERING"
+  address_type  = "INTERNAL"
+  prefix_length = 16
+  network       = "${data.google_compute_network.cloudsql_private_network.self_link}"
+  depends_on    = ["google_project_service.service_networking"]
+}
+
+resource "google_service_networking_connection" "private_vpc_connection" {
+  count                   = var.cloudsql_private ? 1 : 0
+  network                 = "${data.google_compute_network.cloudsql_private_network.self_link}"
+  service                 = "servicenetworking.googleapis.com"
+  reserved_peering_ranges = ["${google_compute_global_address.private_ip_address[count.index].name}"]
```

Comment:
What is the [count.index] used for here?

Reply:
This refers to the "google_compute_global_address" resource.  If var.cloudsql_private is false, then neither this nor that google_compute_global_address resource would be deployed.  If it is true, then that address would be added to the list.

Long story short this is a side-effect of using count.

https://github.com/forseti-security/terraform-google-forseti/blob/1d2a79c3d43b6b285863dcc7595e8f0a1ec14711/modules/cloudsql/main.tf#L42-L53